File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.java
class Solution {
    int mod = (int)1e9+7;
    public int numFactoredBinaryTrees(int[] arr) {
        Arrays.sort(arr);
        int ans =0;
        HashMap<Integer, Long> map = new HashMap<>();

        for(int x:arr)
        {
            long ways =1;
            int max = (int)Math.sqrt(x);

            for(int j = 0,left=arr[0];left<=max;left=arr[++j])
            {
                if(x%left!=0)   continue;
                int right =x/left;
                if(map.containsKey(right))
                {
                    ways = (ways+map.get(left)*map.get(right)*(left==right?1:2))%mod;
                }
            }
            map.put(x,ways);
            ans=(int)(ans+ways)%mod;
        }
        return ans;
    }
}

File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var numFactoredBinaryTrees = function(arr) {
    arr.sort((a,b)=>a-b);
    let map = new Map();
    let mod = 1000000007;
    let res = 0;
    for(let i=0;i<arr.length;i++){
        let count = 1;
        for(let j=0;j<i;j++){
            if(arr[i]%arr[j]==0 && map.has(arr[i]/arr[j])){
                count += map.get(arr[j])*map.get(arr[i]/arr[j]);
            }
        }
        map.set(arr[i],count);
        res += count;
    }
    return res%mod;
    
};